### ISSUE TITLE
Python 3.7 requires pybind11 2.3 (at least on OS X)

### ISSUE THREAD

**Issue by vancleve** [molpopgen/fwdpy11#122]
Compiled fwdpy11:dev_015 (365b839313557da0bfd2e87e37b2e34c49a2999b) on my MacBook Pro running OS 10.13.6 (`Apple LLVM version 9.1.0 (clang-902.0.39.2)`) and `pybind11 2.2.3` using 

- `python3 setup.py build_ext -i` 
- `pip install -e .` 

in source directory. 

The test `python3 tests/test_GSLerror.py` returns 
```
Ran 1 test in 0.001s

OK
```
but 

- `python3 tests/test_fwdpp_types.py`
- `python3 tests/test_SlocusPop.py`
- `python3 tests/test_stateful_fitness.py` 

all hang with minimal CPU activity from the python process. 

I repeated this process with `gcc 7.3.0` and `gcc 8.1.0` from homebrew and obtained the same results (test_GSLerror.py works but the others hang).

Repeating this process on my server with Ubuntu 18.04.1 LTS and `gcc 7.3.0` yields `OK` for all four tests.

**Comment by molpopgen** [comment]
What if you run the tests with the following command (which is the expected one):

```
python -m unittest discover tests
```

The above would typically be run after a local build:

```
python setup.py build_ext -i
```

**Comment by molpopgen** [comment]
I cannot replicate this on my OS X system (Sierra/10.12.6, and conda with Python3 and gcc 4.8.5).

Commands:

```
git checkout dev_015
git submodule update
python setup.py build_ext -i --gcc
python -m unittest discover tests
```

All tests compiled and executed OK.

Unfortunately, I do not have access to a newer OS X system, but I'm guessing that something is funky with homebrew.  Are you sure that it compiled with GCC and not clang?  Typically, you'd have to modify the build command to be something like:

```
CXX=g++-8 CC=gcc-8 python setup.py build_ext -i --gcc
```

Note that you must pass `--gcc` when using GCC on OS X (See README). Otherwise, things get linked to the clang C++ runtime library, which is bad.

I may consider reinstating Travis CI for OS X, but am hesitant to do so because it really slows down the development cycle.  The OS X build bots have a much longer waiting time than do the Linux containers.

**Comment by vancleve** [comment]
I did pass `--gcc`, but did it right after the `setup.py`: 
`CXX=g++-8 CC=gcc-8 python setup.py --gcc build_ext -i`. 

I'll try that now.

**Comment by molpopgen** [comment]
> CXX=g++-8 CC=gcc-8 python setup.py --gcc build_ext -i

That should be ok.  Also remember to clean up when you attempt to recompile:

```
rm -rf build fwdpy11/*.so
```

Python's build system isn't able to do that for you.

**Comment by molpopgen** [comment]
Also, if you are going to run unit tests individually, it is safer to force python to load the local build:

```
PYTHONPATH=. python3 tests/test.py
```

You do not need to do that if you run the `unittest discover` method.

**Comment by vancleve** [comment]
Seems this is a python 3.7 issue. python 3.6.5 works fine.

**Comment by molpopgen** [comment]
> Seems this is a python 3.7 issue. python 3.6.5 works fine.

What Python was your Ubuntu setup using?  I will check python 3.7 from Anaconda on Linux later today and report back here.  Are you able to test Py3.7 from homebrew on Linux?

**Comment by molpopgen** [comment]
Python3 with the Pop OS built-in compiler (gcc 7.3.0) work fine.  I ran several of the tests manually that hung for you, and they worked fine.

Will revisit later with "new" and "old" Anaconda compilers.

**Comment by molpopgen** [comment]
On Linux (Pop OS, aka \approx Ubuntu 18.04), Python 3.7 plus the native GCC 7.3.0 appear to work fine.

I was able to manually run the tests that hung for you and they did not hang for me.

So, this seems to narrow it down to an OS X/home brew interaction, possibly with compiler-related complications.

I will add a set of anaconda + py3.7 tests to the Travis CI matrix for both Linux and OS X to see what happens.

**Comment by molpopgen** [comment]
The branch where I'm adding more Travis CI cases is passing with Py3.7/Anaconda on Linux.  It is failing on OS X, but not due to a hang (which would lead to a timeout error).  Rather, an exception is propagated due to Py_ERROR already being set, which means that something funny is going on handling an exception.  I will attempt to diagnose on my office iMac (OS X Sierra).  But it is looking like the issues on Homebrew are a different thing altogether.

**Comment by molpopgen** [comment]
OK, I am able to replicate the issue with unit tests hanging under py 3.7 on OS X using Anaconda and gcc 4.8.5.  This is on OS X Sierra.

**Comment by molpopgen** [comment]
Solved: with Python 3.7, pybind11 2.3 is required.  This version of pybind11 is unreleased/under development.  

Installing pybind11 from their current git/master solves all of the above issues (on OS X)

```
git clone https://github.com/pybind/pybind11
cd pybind11
pip install .
```

No idea why this matters on OS X but apparently not on Linux.

Installing pybind11/master also removes a ton of (new) compiler warnings. It appears that the py3.7 C API has undergone some changes and pybind11 is in the process of accounting for them now.
